### PR TITLE
Add TLS specification

### PR DESCRIPTION
**Pull request by jonfoster**
Issue #112

### CODE REVIEW

File: specs/tls.rst [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+The *connection scope* information passed in ``scope`` contains an
+``"extensions"`` key, which contains a dictionary of extensions.  Inside that
+dictionary, the key ``"tls"`` identifies the extension specified in this
+document.  The value will be a dictionary with the following entries:
+
+* ``server_cert`` (*Optional[Unicode string]*) -- The PEM-encoded conversion
+  of the x509 certificate sent by the server when establishing the TLS
+  connection.  Some web server implementations may be unable to provide this
+  (e.g. if TLS is terminated by a separate proxy or load balancer); in that
+  case this shall be ``None``.  Mandatory.
```
Comment (andrewgodwin) [inline specs/tls.rst:58]: The way the types are written here is a little confusing if you're not familiar with the typing module - the sentence has both `Optional` and `Mandatory` in it, and the type value is not `Optional[unicode]`. May I suggest the type is instead written as `Unicode string or None` in these cases?

File: specs/tls.rst [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+Rationale (Informative)
+-----------------------
+
+This section explains the choices that led to this specification.
+
+Providing the entire TLS certificates in ``client_cert_chain``, rather than a
+parsed subset:
+
+* makes it easier for webservers to implement, as they do not have to
```
Comment (andrewgodwin) [inline specs/tls.rst:125]: Nit: Let's capitalise the start of each bullet point

File: specs/tls.rst [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+``client_cert_chain``.  However, many ASGI applications will probably find
+that information is sufficient for their application - it provides a simple
+string that identifies the user.  It is simpler to use than parsing the x509
+certificate.  For the server, this information is readily available.
+
+There are theoretical interoperability problems with ``client_cert_name``,
+since it depends on a list of object ID names that is maintained by IANA and
+theoretically can change.  In practise, this is not a real problem, since the
+object IDs that are actually used in certificates have not changed in many
+years.  So in practise it will be fine.
```
Comment (andrewgodwin) [inline specs/tls.rst:195]: practise -> practice

**Review by andrewgodwin — COMMENTED** [review]
I think this looks good overall - left a few small comments, but I do generally like the design and I think it addresses most of the problems raised in the original TLS ticket.

Currently, extensions are in a separate extensions section of the docs; however, this is definitely lengthly enough to need its own one. I think I am OK leaving it in specs for now, but if you could include a cross-link from the extensions section, that'd be great.

File: specs/tls.rst (hunk 3 above)
Comment (jonfoster) [reply specs/tls.rst]: Thanks, should be fixed now.  I ran spellcheck this time, too!

File: specs/tls.rst (hunk 2 above)
Comment (jonfoster) [reply specs/tls.rst]: Done!

File: specs/tls.rst (hunk 1 above)
Comment (jonfoster) [reply specs/tls.rst:58]: Sure, changed.

**Comment by andrewgodwin** [comment]
I'm honestly not sure if we ended up with me having to re-review it or more changes needed. I'll take a look at it sometime this month.

**Comment by andrewgodwin** [comment]
Well, it took longer than a month, but on further review it looks good, so it's merged.

**Pull request merged**